File: apps/runtime-go/internal/contracts/contracts.go
// Package contracts validates documents against the shared JSON Schema
// contracts in packages/contracts/schemas.
//
// The schemas are embedded from gen/schemas, which codegen copies verbatim
// from the source of truth. Embedding the copy rather than reading the
// originals at runtime means the binary carries the exact schemas its
// generated types were built from, and the CI drift gate is what keeps the
// copy honest.
//
// House rule this serves: validate at every boundary. A trace event is checked
// before it goes on the wire, a question request before it is acted on.
package contracts

import (
	"bytes"
	"embed"
	"fmt"
	"io/fs"
	"path"
	"strings"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

//go:embed gen/schemas/*.schema.json
var schemaFS embed.FS

// Name identifies a contract by its discriminator, e.g. "trace_event.v1".
type Name string

const (
	QuestionRequestV1 Name = "question_request.v1"
	TraceEventV1      Name = "trace_event.v1"
	SQLPlanV1         Name = "sql_plan.v1"
	CostLedgerV1      Name = "cost_ledger.v1"
	EvalResultV1      Name = "eval_result.v1"
)

const schemaDir = "gen/schemas"

// compiled holds one validator per contract, built once at init. Compilation
// is not cheap and the schema set is fixed at build time, so doing it per call
// would be pure waste on the request path.
var compiled = mustCompileAll()

// ErrUnknownContract is returned when a name has no embedded schema.
type ErrUnknownContract struct{ Name Name }

func (e ErrUnknownContract) Error() string {
	return fmt.Sprintf("contracts: unknown contract %q", e.Name)
}

// ValidationError reports a document that failed its contract.
type ValidationError struct {
	Name   Name
	Detail string
}

func (e ValidationError) Error() string {
	return fmt.Sprintf("contracts: %s: %s", e.Name, e.Detail)
}

// SafeMessage is what may be shown to a client.
//
// Error() quotes the offending instance values, which for a failed LLM or
// driver response means echoing untrusted — potentially sensitive — content
// straight back to the browser. Handlers log Error() and return this.
func (e ValidationError) SafeMessage() string {
	return fmt.Sprintf("contracts: %s: document failed validation", e.Name)
}

// Names returns every contract this build knows about, for tests and
// diagnostics. Order is not guaranteed.
func Names() []Name {
	out := make([]Name, 0, len(compiled))
	for n := range compiled {
		out = append(out, n)
	}
	return out
}

// Validate checks raw JSON against the named contract.
//
// It returns ErrUnknownContract for a name with no schema and ValidationError
// for a document that does not conform. Both are ordinary errors: callers at a
// boundary should reject the input, not panic.
func Validate(name Name, raw []byte) error {
	sch, ok := compiled[name]
	if !ok {
		return ErrUnknownContract{Name: name}
	}
	doc, err := jsonschema.UnmarshalJSON(bytes.NewReader(raw))
	if err != nil {
		return ValidationError{Name: name, Detail: fmt.Sprintf("malformed JSON: %v", err)}
	}
	if err := sch.Validate(doc); err != nil {
		return ValidationError{Name: name, Detail: err.Error()}
	}
	return nil
}

func mustCompileAll() map[Name]*jsonschema.Schema {
	out, err := compileAll()
	if err != nil {
		// A build whose embedded schemas do not compile is broken in a way no
		// caller can recover from, and failing at init surfaces it on the
		// first test rather than on the first request.
		panic(fmt.Sprintf("contracts: compiling embedded schemas: %v", err))
	}
	return out
}

func compileAll() (map[Name]*jsonschema.Schema, error) {
	entries, err := fs.ReadDir(schemaFS, schemaDir)
	if err != nil {
		return nil, fmt.Errorf("reading embedded schema dir: %w", err)
	}

	c := jsonschema.NewCompiler()
	// Formats are assertions here, not annotations. All three language suites
	// run the same fixture corpus, so they must agree on whether "last tuesday"
	// is a date-time. Leaving formats as annotations would make the Go suite
	// accept fixtures the Python and TypeScript suites reject.
	c.AssertFormat()

	names := make([]Name, 0, len(entries))
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".schema.json") {
			continue
		}
		p := path.Join(schemaDir, e.Name())
		raw, err := schemaFS.ReadFile(p)
		if err != nil {
			return nil, fmt.Errorf("reading %s: %w", p, err)
		}
		doc, err := jsonschema.UnmarshalJSON(bytes.NewReader(raw))
		if err != nil {
			return nil, fmt.Errorf("parsing %s: %w", p, err)
		}
		if err := c.AddResource(e.Name(), doc); err != nil {
			return nil, fmt.Errorf("adding %s: %w", p, err)
		}
		names = append(names, Name(strings.TrimSuffix(e.Name(), ".schema.json")))
	}

	if len(names) == 0 {
		return nil, fmt.Errorf("no schemas embedded from %s", schemaDir)
	}

	out := make(map[Name]*jsonschema.Schema, len(names))
	for _, n := range names {
		sch, err := c.Compile(string(n) + ".schema.json")
		if err != nil {
			return nil, fmt.Errorf("compiling %s: %w", n, err)
		}
		out[n] = sch
	}
	return out, nil
}

File: apps/retriever-python/src/contracts.py
"""Validate documents against the shared JSON Schema contracts.

The mirror of the Go runtime's internal/contracts package, and it works the
same way: the schemas come from src/models/gen/schemas, which codegen copies
verbatim from packages/contracts/schemas, so this service validates against
exactly the files its generated models were built from. The CI drift gate is
what keeps that copy honest.

House rule this serves: validate at every boundary — in on the way in, out on
the way out.
"""

from __future__ import annotations

import json
from functools import cache
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator
from jsonschema import ValidationError as JsonSchemaValidationError

SCHEMA_DIR = Path(__file__).resolve().parent / "models" / "gen" / "schemas"

SCHEMA_SUFFIX = ".schema.json"


class ContractError(Exception):
    """Base for contract failures."""


class UnknownContractError(ContractError):
    """No embedded schema exists for the requested name."""


class DocumentInvalidError(ContractError):
    """A document did not conform to its contract.

    ``detail`` quotes the offending value and is for the log. ``safe_message``
    is what may cross a service boundary: a failing document is often model or
    driver output, and echoing it onward is how untrusted content ends up in a
    browser.
    """

    def __init__(self, name: str, detail: str) -> None:
        super().__init__(f"{name}: {detail}")
        self.name = name
        self.detail = detail

    @property
    def safe_message(self) -> str:
        return f"{self.name}: document failed validation"


@cache
def _validator(name: str) -> Draft202012Validator:
    path = SCHEMA_DIR / f"{name}{SCHEMA_SUFFIX}"
    if not path.is_file():
        raise UnknownContractError(f"unknown contract {name!r}")
    schema = json.loads(path.read_text())
    # format is an assertion, not an annotation. All three language suites run
    # the same fixture corpus, so they have to agree on whether "last tuesday"
    # is a date-time; leaving formats as annotations would make this service
    # accept documents the Go runtime rejects.
    return Draft202012Validator(schema, format_checker=Draft202012Validator.FORMAT_CHECKER)


def contract_names() -> list[str]:
    """Every contract this build can validate."""
    return sorted(p.name.removesuffix(SCHEMA_SUFFIX) for p in SCHEMA_DIR.glob(f"*{SCHEMA_SUFFIX}"))


def validate_document(name: str, document: Any) -> None:
    """Check a document against a named contract.

    Raises:
        UnknownContractError: the name has no embedded schema.
        DocumentInvalidError: the document does not conform.
    """
    validator = _validator(name)
    try:
        validator.validate(document)
    except JsonSchemaValidationError as exc:
        raise DocumentInvalidError(name, exc.message) from exc
